libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -34,21 +37,45 @@ internal class RouteRefreshController(
     private val localDateProvider: () -> Date
 ) {
 
+    @RouteRefreshExtra.RouteRefreshState
+    private var state: String = RouteRefreshExtra.REFRESH_STATE_IDLE
+        set(value) {
+            field = value
+            observer?.onNewState(value)
+        }
+
+    @Volatile
+    internal var observer: RouteRefreshStatesObserver? = null
+        set(value) {
+            field = value
+            field?.onNewState(state)
+        }
+
     internal companion object {
         internal const val LOG_CATEGORY = "RouteRefreshController"
         private const val FAILED_ATTEMPTS_TO_INVALIDATE_EXPIRING_DATA = 3
     }
 
     suspend fun refresh(routes: List<NavigationRoute>): RefreshedRouteInfo {
-        return if (routes.isNotEmpty()) {
-            val routesValidationResults = routes.map { validateRoute(it) }
-            if (routesValidationResults.any { it is RouteValidationResult.Valid }) {
-                tryRefreshingRoutesUntilRouteChanges(routes)
+        try {
```

Comment:
For me both make sense. On the other hand, I would avoid overcomplicated as we're going to migrate to NN and we should keep it as simple as possible

Reply:
> make RouteRefreshController stateful by checking in the beginning if other refresh is running.

This one looks quite simple. I'd implement at least some solution in case migration to NN gets stuck or gets reverted at some point for some reason. If we leave as-is, we risk exposing this inconsistency and forgetting about it.

Reply:
I see other corner cases which also should be checked under the hood like refresh is keep running if any of refresh sets are running. Or what to do if one of refreshes are expired and other is refreshed. What stat should be in this case. I want to suggest change the API that we can refresh only one set of route at time. It covers current needs and not overcomlicated implementation for cases that we haven't had. In a separate PR

Reply:
@VysotskiVadim @dzinad please let me know what do you think, I can cut a ticket for that

Reply:
>  I want to suggest change the API that we can refresh only one set of route at time.

I like the idea. The reason I raised this question is because I worry that on new route set, there is a chance that coroutine won't be cancelled immediately. It's not a problem when we rely only on result, but it can be a problem when the controller calls callbacks with state.

Reply:
But maybe, as there is only 1 UI thread, the coroutine won't be able to change state while it's cancelling 🤔

Reply:
>But maybe, as there is only 1 UI thread, the coroutine won't be able to change state while it's cancelling 🤔

what do you mean?

Reply:
I looked closer at the code and it seems like we now only refresh one set of route at a time (see `MapboxNavigation#createInternalRoutesObserver`). But yeah, the `RouteRefreshController` API does not limit us so it'd be good to control.

Reply:
cut `#NAVAND-644`